### PR TITLE
Moving extension upload pvc

### PR DESCRIPTION
**Pull request by metalice**
Signed-off-by: Matan Schatzman <mschatzm@redhat.com>

## 📝 Description

Moving extension upload PVC from console to be at our reoo
## 🎥 Demo

TODO in follow up PR's:

- Use shared components form disk modal or other things, if applicable - please suggest!
- Move shared utils or components to shared locations
- Extract upload PVC state logic to a reducer or combined state. 

image

### CODE REVIEW

**Review by avivtur — COMMENTED** [review]
Great work :trophy: 
just a few comments :smile:  
also note this PR: https://github.com/openshift/console/pull/12213
that is hiding the checkbox for attaching the data the VM

File: src/views/cdi-upload-provider/PVCDeleteAlertExtension.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+  const isGolden = goldenPvcs?.find(
+    (goldenPvc) => goldenPvc?.metadata?.name === pvc?.metadata?.name,
+  );
+
+  return (
+    <Alert isInline variant={AlertVariant.warning} title="PVC Delete">
+      <Trans t={t} ns="plugin__kubevirt-plugin">
+        <p>
+          Deleting this PVC will also delete{' '}
```
[the remark this answers is not in the document]
Comment (metalice) [reply src/views/cdi-upload-provider/PVCDeleteAlertExtension.tsx:41]: i wouldnt touch it yet.

File: src/views/cdi-upload-provider/hooks/useBaseImages.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        {},
+      ),
+    ];
+  }, [commonTemplates]);
+
+  const pvcs = useK8sWatchResources<{ [key: string]: V1alpha1PersistentVolumeClaim[] }>(pvcWatches);
+  const pvcValues = Object.values(pvcs);
+  const loaded = pvcValues?.every((value) => value.loaded || !!value.loadError);
+  const loadError = pvcValues.find((value) => value.loadError);
+  const pvcData = pvcValues.reduce<V1alpha1PersistentVolumeClaim[]>(
```
[the remark this answers is not in the document]
Comment (metalice) [reply src/views/cdi-upload-provider/hooks/useBaseImages.ts:65]: right, should have removed it

File: src/views/cdi-upload-provider/hooks/useMultipleAccessReviews.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        setLoading(false);
+        const updatedAllowedArr = values.map((result) => ({
+          resourceAttributes: result.spec.resourceAttributes,
+          allowed: result.status.allowed,
+        }));
+        setAllowedArr(updatedAllowedArr);
+      })
+      .catch((e) => {
+        // eslint-disable-next-line no-console
+        console.warn('SelfSubjectAccessReview failed', e);
```
[the remark this answers is not in the document]
Comment (metalice) [reply src/views/cdi-upload-provider/hooks/useMultipleAccessReviews.ts:34]: hmm... old :) , i will leave it

File: src/views/cdi-upload-provider/utils/utils.tsx [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    value = combinedVal.slice(0, index);
+    unit = combinedVal.slice(index);
+  }
+  return [value, unit];
+};
+
+const splitVersion = (osID: string): number[] =>
+  (osID || '')
+    .split(/\D/)
+    .filter((x) => x)
```
[the remark this answers is not in the document]
Comment (metalice) [reply src/views/cdi-upload-provider/utils/utils.tsx:115]: split can return null or undefined

File: src/views/cdi-upload-provider/utils/utils.tsx [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+      // eslint-disable-next-line no-console
+      console.error(`resourcePath: no model for "${modal?.kind}"`);
+    }
+    return;
+  }
+
+  return resourcePathFromModel(modal, name, namespace);
+};
+
+export const stringValueUnitSplit = (combinedVal) => {
```
[the remark this answers is not in the document]
Comment (metalice) [reply src/views/cdi-upload-provider/utils/utils.tsx:99]: maybe can be, wil check in follow up

**Comment by metalice** [comment]
> Great work 🏆 just a few comments 😄 also note this PR: [openshift/console#12213](https://github.com/openshift/console/pull/12213) that is hiding the checkbox for attaching the data the VM

 what is 3 more lines with this PR? added it :)

File: src/views/cdi-upload-provider/hooks/useCDIUpload.tsx [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import { DataUpload, UploadDataProps } from '../utils/types';
+
+const resource: WatchK8sResource = {
+  kind: CDIConfigModelRef,
+  isList: false,
+  namespaced: false,
+  name: 'config',
+};
+
+const useCDIUpload = (): CDIUploadContextProps => {
```
Comment (upalatucci) [inline src/views/cdi-upload-provider/hooks/useCDIUpload.tsx:19]: Already talked about that.
`useCDIUpload` is also used in the disk modal to upload iso files.
It's in `src/utils/hooks/useCDIUpload` and in the near future, we can think of merging these two hooks.
Posted just as memo

File: src/views/cdi-upload-provider/upload-pvc-form/UploadPVC.tsx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+  const [disableFormSubmit, setDisableFormSubmit] = useState(false);
+  const [fileValue, setFileValue] = useState<File>(null);
+  const [fileName, setFileName] = useState('');
+  const [fileNameExtension, setFileNameExtension] = useState('');
+  const [isFileRejected, setIsFileRejected] = useState(false);
+  const [error, setError] = useState<string>('');
+  const [isAllocating, setIsAllocating] = useState(false);
+  const [dvObj, setDvObj] = useState<V1beta1DataVolume>(null);
+  const [commonTemplates, loadedTemplates, errorTemplates] =
+    useK8sWatchResource<V1Template[]>(templatesResource);
```
Comment (upalatucci) [inline src/views/cdi-upload-provider/upload-pvc-form/UploadPVC.tsx:75]: Sooo much states and we do also `setFileValue` `setFileName` and `setFileNameExtension` etc etc in combination. Using a `useReducer` will make the code cleaner. WDYT? In the near future maybe.

File: src/views/cdi-upload-provider/upload-pvc-form/UploadPVCForm.tsx [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+  const [volumeMode, setVolumeMode] = useState<string>();
+  const [requestSizeValue, setRequestSizeValue] = useState<string>('');
+  const [requestSizeUnit, setRequestSizeUnit] = useState<string>('Gi');
+  const [isGolden, setIsGolden] = useState<boolean>(!!osParam);
+  const [os, setOs] = useState<OperatingSystemRecord>();
+  const [pvcSizeFromTemplate, setPvcSizeFromTemplate] = useState<boolean>(false);
+  const [mountAsCDROM, setMountAsCDROM] = useState<boolean>();
+  const [osImageExists, setOsImageExists] = useState<boolean>(false);
+  const defaultSCName = getDefaultStorageClass(storageClasses)?.metadata?.name;
+  const [applySP, setApplySP] = useState<boolean>(true);
```
Comment (upalatucci) [inline src/views/cdi-upload-provider/upload-pvc-form/UploadPVCForm.tsx:86]: Same thing here

Comment (metalice) [reply src/views/cdi-upload-provider/upload-pvc-form/UploadPVCForm.tsx:86]: same thing as above :)

File: src/views/cdi-upload-provider/hooks/useCDIUpload.tsx (hunk 6 above)
Comment (metalice) [reply src/views/cdi-upload-provider/hooks/useCDIUpload.tsx:19]: yes, should be checked in follow up pr

**Pull request merged**